app/src/main/java/fr/free/nrw/commons/contributions/ContributionBoundaryCallback.java

```diff
@@ -0,0 +1,96 @@
+import fr.free.nrw.commons.di.CommonsApplicationModule;
+import fr.free.nrw.commons.media.MediaClient;
+import io.reactivex.Scheduler;
+import io.reactivex.Single;
+import io.reactivex.disposables.CompositeDisposable;
+import io.reactivex.schedulers.Schedulers;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.inject.Inject;
+import javax.inject.Named;
+import timber.log.Timber;
+
+/**
+ * Class that extends PagedList.BoundaryCallback for contributions list It defines the action that
+ * is triggered for various boundary conditions in the list
+ */
+public class ContributionBoundaryCallback extends PagedList.BoundaryCallback<Contribution> {
+
+  private final ContributionsRepository repository;
+  private final SessionManager sessionManager;
+  private final MediaClient mediaClient;
+  private final CompositeDisposable compositeDisposable;
+  private final Scheduler ioThreadScheduler;
+
+  @Inject
+  public ContributionBoundaryCallback(final ContributionsRepository repository,
+      final SessionManager sessionManager,
+      final MediaClient mediaClient,
+      @Named(CommonsApplicationModule.IO_THREAD) final Scheduler ioThreadScheduler) {
+    this.ioThreadScheduler = ioThreadScheduler;
+    this.repository = repository;
+    this.sessionManager = sessionManager;
+    this.mediaClient = mediaClient;
+    compositeDisposable = new CompositeDisposable();
+  }
+
+  /**
+   * It is triggered when the list has no items User's Contributions are then fetched from the
+   * network
+   */
+  @Override
+  public void onZeroItemsLoaded() {
+    fetchContributions();
+  }
+
+  /**
+   * It is triggered when the user scrolls to the top of the list No action is taken at this point
+   */
+  @Override
+  public void onItemAtFrontLoaded(@NonNull final Contribution itemAtFront) {
+  }
+
+  /**
+   * It is triggered when the user scrolls to the end of the list User's Contributions are then
+   * fetched from the network
+   */
+  @Override
+  public void onItemAtEndLoaded(@NonNull final Contribution itemAtEnd) {
+    fetchContributions();
+  }
+
+  /**
+   * Fetches contributions using the MediaWiki API
+   */
+  public void fetchContributions() {
+    compositeDisposable.add(mediaClient.getMediaListForUser(sessionManager.getUserName())
+        .map(mediaList -> {
+          List<Contribution> contributions = new ArrayList<>();
+          for (final Media media : mediaList) {
+            contributions.add(new Contribution(media, Contribution.STATE_COMPLETED));
+          }
+          return contributions;
+        })
+        .subscribeOn(ioThreadScheduler)
+        .subscribe(this::saveContributionsToDB, error -> {
+          Timber.e("Failed to fetch contributions: %s", error.getMessage());
+        }));
+  }
+
+  /**
+   * Saves the contributions the the local DB
+   */
+  private void saveContributionsToDB(final List<Contribution> contributions) {
+    repository.save(contributions);
```

Comment:
I'll move this discussion to the more recent code.
This does nothing without a subscribe, this is just a Single which represents a task that produces a value or an error.
If the contributions are being saved then it is not by this code

Reply:
@macgills There's just a single usage of this method in our codebase. I checked `ContributionsRepository`, `ContributionsLocalDataSource` and `ContributionDao` too. 

Also, I have been testing on a fresh install. 

Can you demonstrate that a `Single` doesn't execute if the result isn't subscribed?

Reply:
Anyways, I have updated the code to add a subscription. Still, it would be great if you could cite a source or demonstrate the above.

Reply:
This is just an object
![image](https://user-images.githubusercontent.com/3358282/81782750-2f03dc00-94f2-11ea-827b-00c427348723.png)

This is the single executing

![image](https://user-images.githubusercontent.com/3358282/81782838-5195f500-94f2-11ea-9289-1cdb7d9e302b.png)

Or we could check the documentation of [Single](http://reactivex.io/documentation/single.html). This is the behaviour of all the reactive types, you need a subscribe call to actually do/receive something. 

[Tutorial](https://medium.com/@factoryhr/understanding-java-rxjava-for-beginners-5eacb8de12ca) [official examples](https://github.com/ReactiveX/RxJava/wiki/How-To-Use-RxJava)

Reply:
Thanks for the example. Hopefully, the updated code should work as expected.